### PR TITLE
Add emscripten

### PR DESCRIPTION
**Pull request by A2va**
I got the env variables from executing `emsdk activate latest`. It displays this on Windows:
```
Resolving SDK alias 'latest' to '3.1.42'
Resolving SDK version '3.1.42' to 'sdk-releases-9d73bf4bd5b5c9ce6e51be0ed5ce6599fcb28e9e-64bit'
Setting the following tools as active:
   node-16.20.0-64bit
   python-3.9.2-nuget-64bit
   java-8.152-64bit
   releases-9d73bf4bd5b5c9ce6e51be0ed5ce6599fcb28e9e-64bit

Adding directories to PATH:
PATH += D:\compiler\emsdk
PATH += D:\compiler\emsdk\upstream\emscripten

Setting environment variables:
PATH = D:\compiler\emsdk;D:\compiler\emsdk\upstream\emscripten;...
EMSDK_NODE = D:\compiler\emsdk\node\16.20.0_64bit\bin\node.exe
EMSDK_PYTHON = D:\compiler\emsdk\python\3.9.2-nuget_64bit\python.exe
JAVA_HOME = D:\compiler\emsdk\java\8.152_64bit
```

And on Linux:
```
Setting up EMSDK environment (suppress these messages with EMSDK_QUIET=1)
Adding directories to PATH:
PATH += /home/a2va/emsdk
PATH += /home/a2va/emsdk/upstream/emscripten
PATH += /home/a2va/emsdk/node/16.20.0_64bit/bin

Setting environment variables:
PATH = /home/a2va/emsdk:/home/a2va/emsdk/upstream/emscripten:/home/a2va/emsdk/node/16.20.0_64bit/bin:
EMSDK = /home/a2va/emsdk
EMSDK_NODE = /home/a2va/emsdk/node/16.20.0_64bit/bin/node
```

### CODE REVIEW

File: packages/e/emscripten/xmake.lua [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+        os.vrunv(py, {path.join(installdir, "emsdk.py"), "install", version})
+
+        -- activation
+        os.vrunv(py, {path.join(installdir, "emsdk.py"), "activate", version})
+
+        -- setup env
+        local node = os.dirs(path.join(installdir, "node/**"))[1]
+        package:addenv("PATH", node)
+
```
Comment (waruqi) [inline packages/e/emscripten/xmake.lua:30]: please remove empty line

File: packages/e/emscripten/xmake.lua [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        package:addenv("PATH", path.join(installdir, "upstream", "emscripten"))
+        package:addenv("PATH", installdir)
+        package:addenv("EMSDK", installdir)
+
+        local exe = package:is_plat("windows") and ".exe" or ""
+
+        package:addenv("EMSDK_NODE", path.join(node, "bin", "node"..exe))
```
Comment (waruqi) [inline packages/e/emscripten/xmake.lua:37]: please add spaces `. .  `

File: packages/e/emscripten/xmake.lua [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        if package:is_plat("windows") then
+            local python = os.dirs(path.join(installdir, "python/**"))[1]
```
Comment (waruqi) [inline packages/e/emscripten/xmake.lua:39]: and here

File: packages/e/emscripten/xmake.lua [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+            package:addenv("EMSDK_PYTHON", path.join(python, "python"..exe))
```
Comment (waruqi) [inline packages/e/emscripten/xmake.lua:40]: and here

**Comment by A2va** [comment]
I struggle to replace `local java = os.dirs(path.join(installdir, "java/**"))[1]` by find_directory because it must detect a folder that is name in function of the version of java in my case ``8.152_64bit`. Could I use regex in find_directory input ?

**Comment by waruqi** [comment]
> I struggle to replace by find_directory because it must detect a folder that is name in function of the version of java in my case ``8.152_64bit`. Could I use regex in find_directory input ?`local java = os.dirs(path.join(installdir, "java/**"))[1]`

you can use lua patten in find_directory. `find_directory("[%d%.]+_64bit", path.join(installdir, "java", "**"))`

File: packages/e/emscripten/xmake.lua [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    set_description("Emscripten: An LLVM-to-WebAssembly Compiler.")
+    set_license("MIT")
+
+    set_urls("https://github.com/emscripten-core/emsdk.git")
+    add_versions("3.1.42", "c2260b4f28f53b411032de0955a6fe6b6bcf3edd")
+
+    add_deps("python")
+
+    on_install("windows", "macosx", "linux", function (package)
+        import("lib.detect.find_directory")
```
Comment (waruqi) [inline packages/e/emscripten/xmake.lua:14]: please add a newline

File: packages/e/emscripten/xmake.lua [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        package:addenv("PATH", node)
+        package:addenv("PATH", path.join(installdir, "upstream", "emscripten"))
+        package:addenv("PATH", installdir)
+        package:addenv("EMSDK", installdir)
+    
+        local exe = package:is_plat("windows") and ".exe" or ""
+
+        package:addenv("EMSDK_NODE", path.join(node, "node" .. exe))
+        if package:is_plat("windows") then
+            local python = find_directory("*", path.join(installdir, "python"))
```
Comment (waruqi) [inline packages/e/emscripten/xmake.lua:38]: add `if python then`

File: packages/e/emscripten/xmake.lua [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+            package:addenv("EMSDK_PYTHON", path.join(python, "python" .. exe))
+            local java = find_directory("*", path.join(installdir, "java"))
```
Comment (waruqi) [inline packages/e/emscripten/xmake.lua:40]: add `if java then`

**Pull request merged**